model/vm_pool.rb

```diff
@@ -15,7 +15,8 @@ def pick_vm
     DB.transaction do
       # first lock the whole pool in the join table so that no other thread can
       # pick a vm from this pool
-      picked_vm = vms_dataset.for_update.where(display_state: "running").first
+      suitable_vms = vms_dataset.for_update.where(display_state: "running")
+      picked_vm = suitable_vms.reject { |vm| GithubRunner[vm_id: vm.id] }.first
```

Comment:
If I understand correctly, the lock is released on the second line, and we use the `add_unique_constraint :vm_id` database constraint to ensure each vm is selected only once. If this fails, the assumption is that a new vm will be selected in the subsequent run. 

However, `VmPool.pick_vm` is invoked during `GithubRunner.assemble`, which is called at the web endpoint. If a database constraint exception is raised because a vm couldn't be picked again, GitHub won't resend the same webhook. Consequently, we would miss the runner request.

Reply:
Lock will be released once the transaction is either committed or aborted. In this case, GithubRunner.assemble starts the transaction. That way, until we create the GithubRunner entity, the lock is not released. This should be safe. 
The only reason I added the unique constraint is to have a safe guard in case things go south ways and it looks like github runners must have unique VMs anyway.

Reply:
We can attempt to consolidate these two lines into a single query to ensure compatibility with VM uniquness check and resolve the N+1 query issue. For instance, our VM pool size for standard-2 is 38, which, in the worst case, would result in 38 additional queries. This could potentially slow down our webhook endpoint much.

First step, we need to add `github_runner` relation to the vm model than eager load it
```suggestion
      suitable_vms = vms_dataset.for_update.where(display_state: "running").eager(:github_runner)
      picked_vm = suitable_vms.reject { |vm| !vm.nil? }.first
```
Then sequel library has some way to check existence of relation in same query. I didn't check it detail yet.

Reply:
I wrote the second message before seeing your comment; feel free to ignore the latter part of it. However, I believe addressing the issue of the 38 queries might be good.

Reply:
It would be better to add a .join to first line instead of adding github_runner relation to the vm model.

Reply:
handled it with a left join to get the ids of the eligible VMs. Run `for_update` later because `for_update` with left join cannot be used together;
```
E, [2023-10-16T11:13:50.264573 #74149] ERROR -- : PG::FeatureNotSupported: ERROR:  FOR UPDATE cannot be applied to the nullable side of an outer join: SELECT * FROM "vm" LEFT JOIN "github_runner" ON ("github_runner"."vm_id" = "vm"."id") WHERE (("vm"."pool_id" = $1) AND ("github_runner"."vm_id" IS NULL) AND ("vm"."display_state" = $2)) FOR UPDATE; ["5a4650fa-2ab1-8776-8a09-d5784b109c76", "running"]
Sequel::DatabaseError: PG::FeatureNotSupported: ERROR:  FOR UPDATE cannot be applied to the nullable side of an outer join
```

Reply:
What about now?

Reply:
Now it is slightly slower than before, but probably in OK range. We can optimize it in the future if necessary.